From ed08644227ffe3440fbda48bf69ca940c90fc3fb Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 29 Dec 2022 18:59:23 +0100
Subject: [PATCH 37/42] Revert "rtw88: Configure the registers from
 rtw_bf_assoc() outside the RCU lock"

This reverts commit 8f5a05b7da6f70b848832f2bcc259c66aee98a63.

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 drivers/net/wireless/realtek/rtw88/bf.c | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)

diff --git a/drivers/net/wireless/realtek/rtw88/bf.c b/drivers/net/wireless/realtek/rtw88/bf.c
index c827c4a28..038a30b17 100644
--- a/drivers/net/wireless/realtek/rtw88/bf.c
+++ b/drivers/net/wireless/realtek/rtw88/bf.c
@@ -49,23 +49,19 @@ void rtw_bf_assoc(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
 
 	sta = ieee80211_find_sta(vif, bssid);
 	if (!sta) {
-		rcu_read_unlock();
-
 		rtw_warn(rtwdev, "failed to find station entry for bss %pM\n",
 			 bssid);
-		return;
+		goto out_unlock;
 	}
 
 	ic_vht_cap = &hw->wiphy->bands[NL80211_BAND_5GHZ]->vht_cap;
 	vht_cap = &sta->deflink.vht_cap;
 
-	rcu_read_unlock();
-
 	if ((ic_vht_cap->cap & IEEE80211_VHT_CAP_MU_BEAMFORMEE_CAPABLE) &&
 	    (vht_cap->cap & IEEE80211_VHT_CAP_MU_BEAMFORMER_CAPABLE)) {
 		if (bfinfo->bfer_mu_cnt >= chip->bfer_mu_max_num) {
 			rtw_dbg(rtwdev, RTW_DBG_BF, "mu bfer number over limit\n");
-			return;
+			goto out_unlock;
 		}
 
 		ether_addr_copy(bfee->mac_addr, bssid);
@@ -79,7 +75,7 @@ void rtw_bf_assoc(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
 		   (vht_cap->cap & IEEE80211_VHT_CAP_SU_BEAMFORMER_CAPABLE)) {
 		if (bfinfo->bfer_su_cnt >= chip->bfer_su_max_num) {
 			rtw_dbg(rtwdev, RTW_DBG_BF, "su bfer number over limit\n");
-			return;
+			goto out_unlock;
 		}
 
 		sound_dim = vht_cap->cap &
@@ -102,6 +98,9 @@ void rtw_bf_assoc(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
 
 		rtw_chip_config_bfee(rtwdev, rtwvif, bfee, true);
 	}
+
+out_unlock:
+	rcu_read_unlock();
 }
 
 void rtw_bf_init_bfer_entry_mu(struct rtw_dev *rtwdev,
-- 
2.39.0.rc2.1.gbd5df96b79

